executor/mpp_gather.go

```diff
@@ -0,0 +1,331 @@
+	}, nil
+}
+
+// ExchangeSender dispatches data to upstream tasks. That means push mode processing,
+type ExchangeSender struct {
+	plannercore.PhysicalExchangerBase
+
+	tasks        []*mppTask
+	exchangeType tipb.ExchangeType
+}
+
+// ToPB generates the pb structure.
+func (e *ExchangeSender) ToPB(ctx sessionctx.Context, storeType kv.StoreType) (*tipb.Executor, error) {
+	child, err := e.Children()[0].ToPB(ctx, kv.TiFlash)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	encodedTask := make([][]byte, 0, len(e.tasks))
+
+	for _, task := range e.tasks {
+		encodedStr, err := task.ToPB().Marshal()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		encodedTask = append(encodedTask, encodedStr)
+	}
+
+	ecExec := &tipb.ExchangeSender{
+		Tp:              e.exchangeType,
+		EncodedTaskMeta: encodedTask,
+		Child:           child,
+	}
+	executorID := e.ExplainID().String()
+	return &tipb.Executor{
+		Tp:             tipb.ExecType_TypeExchangeSender,
+		ExchangeSender: ecExec,
+		ExecutorId:     &executorID,
+	}, nil
+}
+
+// MPPGather dispatch MPP tasks and read data from root tasks.
+type MPPGather struct {
+	// following fields are construct needed
+	baseExecutor
+	is           infoschema.InfoSchema
+	originalPlan plannercore.PhysicalPlan
+	startTS      uint64
+
+	allocTaskID int64
+	mppReqs     []*kv.MPPDispatchRequest
+
+	respIter distsql.SelectResult
+}
+
+func (e *MPPGather) constructMPPTasksImpl(ctx context.Context, p *planFragment) ([]*mppTask, error) {
+	if p.tableScan.Table.GetPartitionInfo() == nil {
+		return e.constructSinglePhysicalTable(ctx, p.tableScan.Table.ID, p.tableScan.Ranges)
+	}
+	tmp, _ := e.is.TableByID(p.tableScan.Table.ID)
+	tbl := tmp.(table.PartitionedTable)
+	partitions, err := partitionPruning(e.ctx, tbl, p.tableScan.PartitionInfo.PruningConds, p.tableScan.PartitionInfo.PartitionNames, p.tableScan.PartitionInfo.Columns, p.tableScan.PartitionInfo.ColumnNames)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	allTasks := make([]*mppTask, 0)
+	for _, part := range partitions {
+		partTasks, err := e.constructSinglePhysicalTable(ctx, part.GetPhysicalID(), p.tableScan.Ranges)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		allTasks = append(allTasks, partTasks...)
+	}
+	return allTasks, nil
+}
+
+func (e *MPPGather) constructSinglePhysicalTable(ctx context.Context, tableID int64, ranges []*ranger.Range) ([]*mppTask, error) {
+	kvRanges := distsql.TableRangesToKVRanges(tableID, ranges, nil)
+	req := &kv.MPPBuildTasksRequest{KeyRanges: kvRanges}
+	stores, err := e.ctx.GetMPPClient().ConstructMPPTasks(ctx, req)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	tasks := make([]*mppTask, 0, len(stores))
+	for _, store := range stores {
+		e.allocTaskID++
+		tasks = append(tasks, &mppTask{taskInfo: store, id: e.allocTaskID, startTs: e.startTS, tableID: tableID})
+	}
+	return tasks, nil
+}
+
+func (e *MPPGather) getPlanFragments(p plannercore.PhysicalPlan, pf *planFragment) {
```

Comment:
How about decoupling this function from `MPPGather` and making it a static function, for example, `func getPlanFragments(ctx, p, pf)`?
Then we can easily add unit tests for it without creating a new `MPPGather` struct.

Reply:
Good idea